xayn-ai/src/ranker/context.rs

```diff
@@ -0,0 +1,117 @@
+use crate::{data::document_data::CoiComponent, ranker::utils::Id};
+
+/// The context used to calculate a document's score.
+/// https://xainag.atlassian.net/wiki/spaces/M2D/pages/770670607/Production+AI+Workflow#3.2-Context-calculations.
+/// outlines the calculation of positive and negative distance factor.
+pub(crate) struct Context {
+    /// Average positive distance.
+    pos_avg: f32,
+    /// Maximum negative distance.
+    neg_max: f32,
+}
+
+impl Context {
+    pub(crate) fn from_cois(cois: &[(Id, CoiComponent)]) -> Self {
+        let cois_len = cois.len() as f32;
+        let pos_avg = cois.iter().map(|(_, coi)| coi.pos_distance).sum::<f32>() / cois_len;
+        let neg_max = cois
+            .iter()
+            .map(|(_, coi)| coi.neg_distance)
+            .fold(f32::MIN, f32::max); // NOTE f32::max considers NaN as smallest value
+
+        Self { pos_avg, neg_max }
+    }
+
+    /// Calculates score from given positive distance and negative distance.
+    pub(crate) fn calculate_score(&self, pos: f32, neg: f32) -> f32 {
```

Comment:
since pos and neg are assumed to be distances, should we check that they are actually >= 0?

Reply:
I'm not sure if it's worth to add a runtime check here and what we should do if one of them is < 0. We can add to the docstring that both values need to be >= 0.

Reply:
adding some documentation is good. and we can also add some debug assertions.

Reply:
done